thetis/coordsys.py

```diff
@@ -5,17 +5,133 @@
 import firedrake as fd
 import pyproj
 import numpy
+from abc import ABC, abstractmethod
 
-
-UTM_ZONE10 = pyproj.Proj(
-    proj='utm',
-    zone=10,
-    datum='WGS84',
-    units='m',
-    errcheck=True)
 LL_WGS84 = pyproj.Proj(proj='latlong', datum='WGS84', errcheck=True)
 
 
+class MeshCoordSystem(ABC):
+    """
+    Base class for horizontal coordinate systems
+
+    Provides methods for coordinate transformations etc.
+    """
+    @abstractmethod
+    def to_lonlat(self, x, y):
+        """Convert coordinates to latitude and longitude"""
+        pass
+
+    @abstractmethod
+    def get_vector_rotator(self, x, y):
+        """
+        Returns a vector rotator object.
+
+        The rotator converst vector-valued data to/from longitude, latitude
+        coordinates.
+        """
+        pass
+
+
+def proj_transform(x, y, trans=None, source=None, destination=None):
+    """
+    Transform coordinates from source to target system.
+
+    :arg x,y: coordinates, float or numpy.array_like
+    :kwarg trans: pyproj Transformer object (optional)
+    :kwarg source: source coordinate system, Proj object
+    :kwarg destination: destination coordinate system, Proj object
+    """
+    if trans is None:
+        assert source is not None and destination is not None, \
+            'Either trans or source and destination must be defined'
+        trans = None
+    x_is_array = isinstance(x, numpy.ndarray)
+    y_is_array = isinstance(y, numpy.ndarray)
+    numpy_inputs = x_is_array or y_is_array
+    if numpy_inputs:
+        assert x_is_array and y_is_array, 'both x and y must be numpy arrays'
+        assert x.shape == y.shape, 'x and y must have same shape'
+        # NOTE proj may give wrong results if inputs contain NaNs
```

Comment:
Maybe raise this as a warning? It is unlikely that users will see this note in the source code.

Reply:
Yeah this is the justification why we the workaround, i.e. figure out the `good_ix` indices instead of passing `x` and `y` directly to `transform`. I'll update the comment